message view: Allow unmute a topic from its recipient bar.

akshatdalton opened the pull request:
Earlier a user can only mute a topic from its recipient bar but can't unmute it from there. This commit fixes that bug and allows a user to also unmute the topic from its recipient bar.

Fixes #15223.

ryanreh99 commented on static/templates/recipient_row.hbs:
```diff
@@ -50,7 +50,11 @@
                 </span>
 
+                {{#if topic_muted}}
+                <i class="fa fa-bell on_hover_topic_unmute recipient_bar_icon" data-stream-id="{{stream_id}}" data-topic-name="{{topic}}" title="{{t 'Unmute topic' }}" role="button" tabindex="0" aria-label="{{t 'Mute topic' }} (M)"></i>
```
I guess the `aria-label` attribute should be "Unmute" too.
And we should make the same change in `recent_topic_row.hbs`

ryanreh99 reviewed (commented):
I tested this locally and the message feed gets updated fine but the mute button does not get updated.
To view the button change one has to re-narrow again. (click the same topic in the left sidebar or reload).

This should be done through the [events codepath](https://github.com/zulip/zulip/blob/master/static/js/server_events_dispatch.js#L77) instead of the click handler codepath to get real time updates.

Because every clients UI should get updated when the action takes place.
(When a user mutes the topic in one tab, the other tab should also get the UI update.)

timabbott commented:
@akshatdalton can you update this PR given the feedback?

akshatdalton commented:
Yes, working on that.

akshatdalton commented:
To update the mute button on every muting/unmuting of a topic what actually happens that -
* `_rerender_header` function is called in `message_list_view.js` which in turn calls `populate_group_from_message_container` function to re-populate the message group.
* But to re-populate the message group, we need `message_group_id`.

So this `message_group_id` is provided in two ways:
1. If someone clicks the `bell icon` in the recipient bar, then we can easily find the `message_group_id` by providing the `recipient_row`, which is handled [here](https://github.com/zulip/zulip/pull/16871/files#diff-dca917012760f5ad2b7cebeba4864d546765bbc226d279160fa7cd9d773c5610R169-R172).

2. But if someone mutes/unmutes a topic from stream popover, then we don't have `recipient_row`, and we need to find the `message_group` using `stream_id` and `topic` by looking into `message_lists.current.view._message_groups` which is handled [here](https://github.com/zulip/zulip/pull/16871/files#diff-dca917012760f5ad2b7cebeba4864d546765bbc226d279160fa7cd9d773c5610R174-R180).

Following the second point, if a user has two open tabs, one in `recent topics` and another in, say, `Topic 1` and then if a user mutes/unmutes `Topic 1` from the `recent topics` section, then that update in the recipient bar is not visible. This is because we use [this logic here](https://github.com/zulip/zulip/pull/16871/files#diff-dca917012760f5ad2b7cebeba4864d546765bbc226d279160fa7cd9d773c5610R174-R180) which uses `message_groups`, and `message_groups` are updated each time we narrow to a different section. Hence updated `message_group_id`s differ from the original ones. This, in turn, doesn't update the original recipient row (neither affects the other recipient rows).

So the last point is the only problem that I see (as an edge case), but other than that, there is no problem. This works fine for the general use case, i.e. updating the `bell icon` if a user mutes/unmutes a topic from its recipient bar or stream popover.

Other than that, this `push` adds a helper function: `mute_or_unmute_topic` in `click_handlers.js` to mute and unmute a topic.

@timabbott, please review.

akshatdalton commented:
Also, [this](https://github.com/zulip/zulip/blob/master/static/js/stream_popover.js#L236-L238) block of code looks unnecessary.
That can be just handled by:
```.js
const can_mute_topic = muting.is_topic_muted(sub.stream_id, topic_name);
```
And doing an `if`/`else` over [here](https://github.com/zulip/zulip/blob/master/static/templates/topic_sidebar_actions.hbs#L11-L27).

timabbott commented:
I merged the first commit as 1229945add2a875cb217ce3f511a1c23668357d9.  The second commit is progress but still need some work; I'll post some details.

timabbott commented on static/js/muting_ui.js:
```diff
@@ -207,3 +210,20 @@ export function handle_user_updates(muted_user_ids) {
+            message_lists.current.view._rerender_header(message_group.message_containers);
+        }
+    }
+}
```
This function is wrong; and I think the fix involves no longer passing `row` here.  The bugs are (1) that we can have more than one message list rendered at a time that might need updating and (2) we might have multiple recipient bars for a given topic that need updating (E.g. in a search or stream-level narrow where there's interleaved messages from the topic in question).

I think the existing call to `rerender_on_topic_update` should handle what we need here.

timabbott commented:
@akshatdalton I added a few commits working in the direction that I think this needs to go, including some debug commits before running out of time for helping with this.

I think the core structural problem here is that the state change in `muting.js` is somehow not being propagated before we rerender the message feed (which should, in theory, update the header automatically without the new functions you'd added).

(Or maybe we're passing the wrong parameters to `muting.is_topic_muted` and it's e.g. trying to use a string in an integer dictionary or vise versa)

abhijeetbodas2001 commented on static/js/muting.js:
```diff
@@ -36,7 +36,7 @@ export function is_topic_muted(stream_id, topic) {
     }
     const sub_dict = muted_topics.get(stream_id);
-    return sub_dict && sub_dict.get(topic);
+    return (sub_dict && sub_dict.get(topic)) || false;
```
Isn't this redundant code? `flag OR false` is identically equal to `flag`.

akshatdalton replied:
The only possible cases here are:

1. `undefined` || `false` => `false`
2. `some integer` || `false` => `integer itself`.
(here it can also be `true` || `false` => `true`).

abhijeetbodas2001 replied:
Is the intent to convert the return value to a Boolean? Maybe `return (sub_dict && sub_dict.get(topic)) !== undefined` might be cleaner then?

abhijeetbodas2001 commented on static/js/message_list.js:
```diff
@@ -384,6 +384,15 @@ export class MessageList {
+        // Hence, we rerender them by calling the rerender function
+        // for muted topics (which is done just before we update the
+        // items for topic muting).
+        this.rerender();
         this.data.update_items_for_topic_muting();
         if (this.data.excludes_muted_topics) {
             this.rerender();
```
Why not just remove the condition? (But please make sure we rerender *after* updating the data)
I also did not understand your comment. Removing `this.rerender()` out of the condition is essential here because otherwise the bell icon will not update in stream+topic narrows, where `excludes_muted_topics` is set as `false`. Is there any other reason too?

akshatdalton replied:
**Note:** This added logic works in a similar fashion to how we update the topic name in the recipient bar.

`rerender` function in `message_list.js` ultimately calls `render` function in `rerender_with_target_scrolltop` function inside `message_list_view.js`. So there you will observe that we render for `messages`(or `message_containers`) passed in the function and then we build `new_message_groups` there. So 2 things are important here the `messages` that we pass and the `new_message_groups` that are formed. Before we have called `this.data.update_items_for_topic_muting` which updates `this._items` (or `all_messages()`), it is necessary to `rerender` for the messages (which we have after muting/unmuting the topic). And `data.update_items_for_topic_muting` updates the `this._items` list (to store only `unmuted_messages`) to rerender them (or collapsing the view for muted topics).
After we have new message groups, it won't be possible to `populate_group_from_message_container` for the **original** message container (Sorry but this point will be a little complex to understand unless you look in the codebase).
`message_groups` differ for `current` and `home` `message_list`.

So, the 2nd call is important for the `unmuted_messages` to rerender them (or collapse the muted messages). Because we don't have them unless we call `data.update_items_for_topic_muting` and once we call `data.update_items_for_topic_muting` we don't have muted message in the `current`/`home` `message_list` and hence the first call is important in this way.
See the origin: https://github.com/zulip/zulip/commit/660475bd0cda520eea86a0e3bcfeedc5245ed352

I hope this clears.

abhijeetbodas2001 replied:
I understood your explanation, but not the reasoning behind why we need to care about the *original* (ie, pre-muting) messages/message groups/etc at all?
Once you update the set of messages in the `data`, the new groups will be created based on these new messages right?

It will be really helpful if you can mention what *user-facing* bug rerendering twice prevents.

abhijeetbodas2001 commented on static/js/muting_ui.js:
```diff
@@ -65,19 +65,21 @@ export function mute_topic(stream_id, topic) {
-            });
+            if (hotkey_used) {
+                feedback_widget.show({
+                    populate(container) {
```
An earlier return might be cleaner, but it isn't a big deal

```javascript
if (!hotkey_used) {
    return;
}

// Create feedback widget
```

abhijeetbodas2001 reviewed (commented):
Thanks a lot for working on this! I left some comments and also tested this out. Works correctly everywhere except in stream+topic narrows, where the bell icon does not update. But I think that should be resolved if you just remove the `excludes_muted_topics` check in the message list function.

akshatdalton commented:
In the 1st commit, I've fixed the muting/unmuting of a topic from its recipient bar. Added related comment in the code and similarly updated the commit message. **Please see: https://github.com/zulip/zulip/pull/16871#discussion_r623785300**.
2nd commit just removes the unnecessary extra code.
3rd commit restricts the use of confirmation pop-up on muting the topic when the hotkey: `M` is used.

akshatdalton commented:
Okay, in this push, I've updated the comment in `message_list.js` (**please see: https://github.com/zulip/zulip/pull/16871#discussion_r623785300**).

In the updated comment, I've tried my best to clarify what is happening and why the two calls for `this.rerender` are important.
I hope this clears.

timabbott commented:
OK -- I think the key thing that was missing in this explanation was just the simple fact that we need to rerender recipient rows even if `!excludes_topic_mutes`.  The conclusion is that we need to remove the case where we didn't rerender; there's no reason to rerender twice -- we can just remove that if statement.  

I merged this after a bundle of changes:
* In 91afcda484f3cf7272168dfd11140a8237884bfd, wrote nicer comments, changed the parameter name, and used the early return pattern.
* In the main commit for #15223, changed `message_list.js` to do an unconditional rerender, and added comments about a potential path to avoiding rerenders in cases where we can.
* Also adjusted the styling and iconography to make the on-hover recipient row UI feel more intuitive.

akshatdalton commented:
I'm sorry about not providing solid information about the things going on. After reading your comments, I realised I was only putting the information about the changes which happen to rerender the recipient bar but not explaining the fact you have added.
I have gone through the changes you've made, and I'll remember these for the next time -- about the explanation part to be precise.
